### PR TITLE
Aj400

### PR DESCRIPTION
**Pull request by calypsomatic**
AJ-400 
Creates a helper trait with functions for failing out of tests that use eventually more quickly and puts them in use in a single test file to begin with.

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+    * Wraps a function in an eventually wrapped in a try-catch;
+    * Functions as an eventually but will fail test without retrying if test-ending errors thrown in the test function
+    * @param timeout
+    * @param interval
+    * @param testFunction Test to run; should include a failAndEscape (could be within assertWithEscape)
+    * @tparam T the result of the testFunction
+    * @return The result of the testFunction (an assertion) if an error is not thrown
+    */
+  def eventuallyWithEscape[T](timeout: Timeout, interval: Interval, testFunction: => T): T =
```
Comment (jordansafer) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: nit: I'd vote to put the timeout/interval arguments second, and then make them optional. Something like:
`def eventuallyWithEscape[T](testFunction: => T, timeout: Timout=None, interval=None): T =`
And then combine the two functions. Maybe there's an even fancier technique out there to mirror the eventually functions? (Is it possible to have a curried function where the first parameter is either the timeout or the function? I think this would be more work than it's worth)

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala [hunk 2]
```diff
@@ -67,11 +68,16 @@ class AsyncImportSpec
...
+            val importStatus = blockForStringBody(resp).parseJson.asJsObject.fields.get("status").value
+            assertWithEscape(
+               importStatus shouldBe JsString("Done"),
+              importStatus.equals(JsString("Error")),
+              "Error encountered during import, failing test early"
+            )
+          }
 
             // inspect data entities and confirm correct import as owner
             eventually {
```
Comment (jordansafer) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala:82]: Should we migrate this eventually call too?

**Review by jordansafer — APPROVED** [review]
Looks good!!

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala (hunk 1 above)
Comment (calypsomatic) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: In the original eventually they're implicit https://github.com/scalatest/scalatest/blob/7aeda262e249a66d2319b6b2f5daa1b7d0cf9a4f/jvm/core/src/main/scala/org/scalatest/concurrent/Eventually.scala#L414
so that if you don't supply them it grabs them from PatienceConfig.  So I was relying on the `eventually` inside `eventuallyWithEscape` to take care of this in the second function that only has one parameter.  I don't quite understand how implicit works so I didn't want to try doing that myself, but I'm happy to rewrite/rearrange these functions however works best.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala (hunk 2 above)
Comment (calypsomatic) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala:82]: For this one and a few others it wasn't obvious to me what a fail condition would be.  I just updated the ones that are checking an import status, since it's easy to know that "Error" means it's not ever going to get to done.  This one is checking all the contents of a map.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala [hunk 3]
```diff
@@ -67,11 +68,16 @@ class AsyncImportSpec
...
+            eventuallyWithEscape{
+            val resp: HttpResponse = Orchestration.getRequest( s"${workspaceUrl(projectName, workspaceName)}/importPFB/$importJobId")
+            resp.status shouldBe StatusCodes.OK
+            val importStatus = blockForStringBody(resp).parseJson.asJsObject.fields.get("status").value
+            assertWithEscape(
+               importStatus shouldBe JsString("Done"),
+              importStatus.equals(JsString("Error")),
+              "Error encountered during import, failing test early"
+            )
+          }
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala]: nit: indentation/formatting

EDIT: I'm seeing some funky formatting in a few places, maybe give this a `Code -> Reformat Code` in IntelliJ?

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala [hunk 4]
```diff
@@ -244,8 +255,8 @@ class AsyncImportSpec
         respStatus shouldBe StatusCodes.OK
         val importStatus = blockForStringBody(resp).parseJson.asJsObject.fields.get("status").value
         logger.info(s"[$requestId] Import Service job status for import job $importJobId is [$importStatus]")
-        importStatus shouldBe JsString ("Done")
-      }
+        assertWithEscape(importStatus shouldBe JsString ("Done"), importStatus.equals(JsString ("Error")), "Import error occurred, failing test")
+      })
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala]: this changes how the parens are balanced - it moves one close-paren from line 249 down to line 259. See my comment below on how to avoid that …

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala (hunk 1 above)
Comment (jordansafer) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: I see. I would still vote for optional arguments in that case, to reduce duplication. We could just call eventually with the timeout arguments if they're not none, but leave them out otherwise.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala (hunk 2 above)
Comment (jordansafer) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala:82]: Makes sense!

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala [hunk 5]
```diff
@@ -31,6 +31,7 @@ class AsyncImportSpec
   extends AnyFreeSpec
     with Matchers
     with Eventually
+    with AsyncTestUtils
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala:33]: nit: since `AsyncImportSpec` extends `AsyncTestUtils`, and `AsyncTestUtils` extends `Eventually`, you could remove the `Eventually` from `AsyncImportSpec`

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala (hunk 1 above)
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: if you change this signature to be:
```scala
  def eventuallyWithEscape[T](timeout: Timeout, interval: Interval)(testFunction: => T)
```
you can avoid rebalancing the parens in `AsyncImportSpec`. BUT, see the other thread on just copying the method signature from ScalaTest's `eventually`.

Comment (davidangb) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: I think the best thing to do here is to copy the signature of the `eventually` that you're wrapping, which is [this one](https://github.com/scalatest/scalatest/blob/7aeda262e249a66d2319b6b2f5daa1b7d0cf9a4f/jvm/core/src/main/scala/org/scalatest/concurrent/Eventually.scala#L312). Since its signature is
```scala
def eventually[T](timeout: Timeout, interval: Interval)(fun: => T)(implicit retrying: Retrying[T], pos: source.Position): T
```
you'd have:
```scala
def eventuallyWithEscape[T](timeout: Timeout, interval: Interval)(testFunction: => T)(implicit retrying: Retrying[T], pos: source.Position): T
```
( do the same for the `eventuallyWithEscape` at line 40 too )

FYI, IntelliJ's `View -> Show Implicit Hints` and sometimes `View -> Expand Implicit Hints` can be helpful to follow all the sneaky implicits through! Mostly I hate implicits.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    } catch {
+      case ex:FactoryConfigurationError => throw new TestEscapedException(ex.getMessage)
+    }
+
+  /**
+    * Will fail a test that uses eventually instead of retrying
+    * @param message Error message
+    * @return
+    */
+  def failAndEscape(message: String): Nothing = { throw new FactoryConfigurationError(message) }
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: not necessary for this PR, because this is all rather prospective at this point: I could see some point in the future also adding a 
```scala
def failAndEscape(exception: Exception): Nothing = throw new FactoryConfigurationError(exception)
```
method that allows the original exception to propagate. Of course this would need some corresponding support in `eventuallyWithEscape` to check if the `FactoryConfigurationError` contains an exception and rethrow it.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    * in which case it throws a test-failing error
+    * @param assertion Scalatest assertion to be checked for success or failure
+    * @param failCondition Condition that, if met, will end an asynchronous test with failure immediately
+    * @param failMessage Message to report in test failure
+    * @return test assertion
+    */
+  def assertWithEscape(assertion: => Assertion, failCondition: Boolean, failMessage: String): Assertion =
+    if (failCondition) failAndEscape(failMessage)
+    else assertion
+}
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala:81]: does this work if `failCondition` is also a pass-by-name `Assertion`? Like so:
```scala
  def assertWithEscape(assertion: => Assertion, failCondition: => Assertion, failMessage: String): Assertion = {
    try {
      failCondition
    } catch {
      case ex: Exception => failAndEscape(ex)
    }
    assertion
  }
```

then callers could do something like:
```scala
assertWithEscape(
  importStatus shouldBe JsString("Done"),
  importStatus should not be JsString("Error"),
  "Error encountered during import, failing test early")
```

**Review by davidangb — COMMENTED** [review]
nice job on this! Comments inline …

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala (hunk 7 above)
Comment (calypsomatic) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala:81]: I like that idea!

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala [hunk 8]
```diff
@@ -228,42 +238,42 @@ class AsyncImportSpec
...
         val respStatus = resp.status
-        logger.info(s"[$requestId] HTTP response status for import job $importJobId status request is [${respStatus.intValue()}]. Elapsed: ${humanReadableMillis(System.currentTimeMillis()-startTime)}")
+        logger.info(s"[$requestId] HTTP response status for import job $importJobId status request is [${respStatus.intValue()}]. Elapsed: ${humanReadableMillis(System.currentTimeMillis() - startTime)}")
         respStatus shouldBe StatusCodes.OK
         val importStatus = blockForStringBody(resp).parseJson.asJsObject.fields.get("status").value
         logger.info(s"[$requestId] Import Service job status for import job $importJobId is [$importStatus]")
-        importStatus shouldBe JsString ("Done")
-      }
+        assertWithEscape(importStatus shouldBe JsString("Done"), importStatus should not equal JsString("Error"), "Import error occurred, failing test")
+      })
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala]: now that the signatures for `eventuallyWithEscape` mirror those of `eventually`, please revert the changes to parens here - move the close-paren from line 258 to line 248

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    * @param assertion     Scalatest assertion to be checked for success or failure
+    * @param failCondition Condition that, if met, will end an asynchronous test with failure immediately
+    * @param failMessage   Message to report in test failure
+    * @return test assertion
+    */
+  def assertWithEscape(assertion: => Assertion, failCondition: => Assertion, failMessage: String): Assertion = {
+    try {
+      failCondition
+    } catch {
+      case ex: Exception => failAndEscape(failMessage)
```
Comment (davidangb) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncTestUtils.scala]: this will lose all details from the thrown exception; how about:
```scala
case ex: Exception => failAndEscape(s"$failMessage - underlying error message is: ${ex.getMessage}")
```
or something like that?

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala (hunk 8 above)
Comment (calypsomatic) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/AsyncImportSpec.scala]: Whoops! Forgot that part.  Thanks!

**Review by davidangb — APPROVED** [review]
:+1: assuming tests pass!
